### ISSUE TITLE
Using diffrax to integrate system dynamics for Model Predictive Control

### ISSUE THREAD

**Issue by Hs293Go** [patrick-kidger/diffrax#393]
Diffrax docs' [Kalman Filter Example](https://docs.kidger.site/diffrax/examples/kalman_filter/) encouraged me to apply diffrax to other linear-quadratic control problems, such as the following
``` math
\min_{u(t)} x_N^TQx_N + \sum_{k=0}^N x_k^T Q x_k + u_k^T R u_k\quad \text{subject to}\ \dot{x}(t) = f(x(t), u(t)),\ x(0) = x_{init}
```
essentially finding the optimal control $u$ over a prediction horizon $T$ discretized into $N$ segments. 

Diffrax seems like a good fit for solving the IVP $\dot{x}(t) = f(x(t), u(t)),\ x(0) = x_{init}$ over the prediction horizon to compute $x_{0,\ldots,N}$. In this case...

- I would evaluate $f$ at fixed time points so I don't want to pay any overhead of error estimation. (It appears as low-order as `Heun` already has an embedded error estimator)
- The control terms $u_{0, \ldots,  N}$, being my optimized variables, are known beforehand at the discretization points; I don't want pay the overhead of interpolation per `interpolate_us` in the KF example
- My $f$ is nonlinear in $u$ and I'm unsure if `ControlTerm` is useful here

Is diffrax the right tool for this job? Is yes, what is the best approach; Are there any idioms?

**Comment by patrick-kidger** [comment]
Yup, this sounds exactly like what Diffrax is for. You might also find [this example](https://docs.kidger.site/diffrax/examples/forcing/) a useful reference for how to insert the control input `u`.

Touching on your other points:
- If you don't want to pay the computational cost of using an error estimator, then feel free to use any solver + a fixed step size. Their built-in error estimators will actually be compiled out if they're not used.
- The overhead of interpolation is probably best handled by writing your own function
    ```python
    def u(t):
        # lookup values of `u_i` here given `t`. Implement arbitrary behaviour for those `t` that aren't in the discretisation points.
    ```
- You don't want `ControlTerm`; this is used for SDEs etc. (The word "control" here is in the sense of a "controlled differential equation", e.g. the SDE-like objects `dy(t) = f(y(t)) dx(t)`.) You just want an ODE like in the example above.

I hope that helps!

**Comment by mathDR** [comment]
@patrick-kidger  If you wanted to have a random component for a user defined `u(t)`, how would that be accomplished?

In particular if I have a function that I wish to solve `f(t, x, u)` and I input a function `Callable` for `u(t)` as an `args` like
```
def f(t, x, args):
  u_fn = args
  u = u_fn(t, x)
  ...
```
things are fine.  But if `u(t)` has a random component, like:
```
def u(t, x, key):
  return jax.random.uniform(key, shape=())
```
or something, how would that be accomplished?

NOTE:  I have tried appending a key to the state when calling `diffeqsolve` but that doesn't work (see below).

For example
```
def f(t, x_and_key, args):
  u_fn = args
  x, k = x_and_key 
  k, subkey = jax.random.split(k)
  u = u_fn(t, x, k)
  ...
  return (new_x, k)
```
but when I run this, it is converting the type of the key from `uint32` to `float32` since `x` is of type `float`.  I am trying to pin down where/why this happens, but haven't been able to yet (I am using `eqx.filter_jit` on the entire `solve` function, but it persists even if I do not complie).

**Comment by patrick-kidger** [comment]
That last `f` you've defined appears to be trying to use a different key for every vector field evaluation? If you're looking to solve a random ODE then typically you would pick a single key and fix it for the entrire solve, in which case passing it in through `args` should work.

**Comment by mathDR** [comment]
So I recently discovered the `jax.random.fold_in` function.

So now I can pass in the the key as you said in `args` and then "update" it via:
```
this_iteration_key = jax.random.fold_in(
    key, jax.lax.convert_element_type(sim_time * 1000000, jnp.int32)
)
```
